File: src/main/java/org/moinex/ui/dialog/wallettransaction/ChangeWalletTypeController.java
/*
 * Filename: ChangeWalletTypeController.java
 * Created on: October  6, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.wallettransaction;

import jakarta.persistence.EntityNotFoundException;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.entities.wallettransaction.Wallet;
import org.moinex.entities.wallettransaction.WalletType;
import org.moinex.error.MoinexException;
import org.moinex.services.WalletService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Change Wallet Type dialog
 */
@Controller
@NoArgsConstructor
public class ChangeWalletTypeController
{
    @FXML
    private ComboBox<Wallet> walletComboBox;

    @FXML
    private ComboBox<WalletType> newTypeComboBox;

    @FXML
    private Label currentTypeLabel;

    private List<Wallet> wallets;

    private List<WalletType> walletTypes;

    private WalletService walletService;

    private Wallet wallet = null;

    /**
     * Constructor
     * @param walletService WalletService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public ChangeWalletTypeController(WalletService walletService)
    {
        this.walletService = walletService;
    }

    public void setWalletComboBox(Wallet wt)
    {
        if (wallets.stream().noneMatch(w -> w.getId().equals(wt.getId())))
        {
            return;
        }

        this.wallet = wt;
        walletComboBox.setValue(wallet);
        updateCurrentTypeLabel(wallet);
    }

    @FXML
    private void initialize()
    {
        configureComboBoxes();

        loadWalletsFromDatabase();
        loadWalletTypesFromDatabase();

        populateComboBoxes();

        // Set the current type label
        walletComboBox.setOnAction(e -> {
            Wallet wallet = walletComboBox.getValue();
            updateCurrentTypeLabel(wallet);
        });
    }

    @FXML
    private void handleSave()
    {
        Wallet     wallet        = walletComboBox.getValue();
        WalletType walletNewType = newTypeComboBox.getValue();

        if (wallet == null || walletNewType == null)
        {
            WindowUtils.showInformationDialog(
                "Empty fields",
                "Please fill all required fields before saving");
            return;
        }

        try
        {
            walletService.changeWalletType(wallet.getId(), walletNewType);

            WindowUtils.showSuccessDialog("Wallet type changed",
                                          "The wallet type was successfully changed.");
        }
        catch (EntityNotFoundException | MoinexException.AttributeAlreadySetException e)
        {
            WindowUtils.showErrorDialog("Invalid input", e.getMessage());
            return;
        }

        Stage stage = (Stage)walletComboBox.getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleCancel()
    {
        Stage stage = (Stage)walletComboBox.getScene().getWindow();
        stage.close();
    }

    private void loadWalletsFromDatabase()
    {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName();
    }

    private void loadWalletTypesFromDatabase()
    {
        walletTypes = walletService.getAllWalletTypes();
    }

    private void populateComboBoxes()
    {
        walletComboBox.getItems().addAll(wallets);

        String nameToMove = "Others";

        // Move the "Others" wallet type to the end of the list
        if (walletTypes.stream().anyMatch(n -> n.getName().equals(nameToMove)))
        {
            WalletType walletType =
                walletTypes.stream()
                    .filter(wt -> wt.getName().equals(nameToMove))
                    .findFirst()
                    .orElseThrow(
                        () -> new IllegalStateException("Invalid wallet type"));

            walletTypes.remove(walletType);
            walletTypes.add(walletType);
        }

        // Add wallet types, but remove the default goal wallet type
        newTypeComboBox.getItems().addAll(
            walletTypes.stream()
                .filter(
                    wt -> !wt.getName().equals(Constants.GOAL_DEFAULT_WALLET_TYPE_NAME))
                .toList());
    }

    private void configureComboBoxes()
    {
        UIUtils.configureComboBox(walletComboBox, Wallet::getName);
        UIUtils.configureComboBox(newTypeComboBox, WalletType::getName);
    }

    private void updateCurrentTypeLabel(Wallet wt)
    {
        if (wt == null)
        {
            currentTypeLabel.setText("-");
            return;
        }

        currentTypeLabel.setText(wt.getType().getName());
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/update/ChangeWalletTypeController.kt
/*
 * Filename: ChangeWalletTypeController.kt (original filename: ChangeWalletTypeController.java)
 * Created on: October  6, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.Label
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.exception.MoinexException
import org.moinex.model.wallettransaction.Wallet
import org.moinex.model.wallettransaction.WalletType
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.springframework.stereotype.Controller

@Controller
class ChangeWalletTypeController(
    private val walletService: WalletService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var walletComboBox: ComboBox<Wallet>

    @FXML
    private lateinit var newTypeComboBox: ComboBox<WalletType>

    @FXML
    private lateinit var currentTypeLabel: Label

    private var wallets: List<Wallet> = emptyList()
    private var walletTypes: List<WalletType> = emptyList()

    companion object {
        private const val OTHER_WALLET_TYPE_NAME = "Others"
    }

    fun setWalletComboBox(wt: Wallet) {
        if (wallets.none { it.id == wt.id }) {
            return
        }
        walletComboBox.value = wt
        updateCurrentTypeLabel(wt)
    }

    @FXML
    private fun initialize() {
        configureComboBoxes()
        loadWalletsFromDatabase()
        loadWalletTypesFromDatabase()
        populateComboBoxes()

        walletComboBox.setOnAction {
            val wt = walletComboBox.value
            updateCurrentTypeLabel(wt)
        }
    }

    @FXML
    private fun handleSave() {
        val wt = walletComboBox.value
        val walletNewType = newTypeComboBox.value

        if (wt == null || walletNewType == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            wt.type = walletNewType
            walletService.changeWalletType(wt)

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_WALLET_TYPE_CHANGED_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_WALLET_TYPE_CHANGED_MESSAGE),
            )
        }.onFailure { e ->
            when (e) {
                is EntityNotFoundException, is MoinexException.AttributeAlreadySetException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_INPUT_TITLE,
                        ),
                        e.message ?: "Unknown error",
                    )
                    return
                }
                else -> throw e
            }
        }

        val stage = walletComboBox.scene.window as Stage
        stage.close()
    }

    @FXML
    private fun handleCancel() {
        val stage = walletComboBox.scene.window as Stage
        stage.close()
    }

    private fun loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName()
    }

    private fun loadWalletTypesFromDatabase() {
        walletTypes = walletService.getAllWalletTypes()
    }

    private fun populateComboBoxes() {
        walletComboBox.items.addAll(wallets)

        val mutableTypes = walletTypes.toMutableList()

        mutableTypes.find { it.name == OTHER_WALLET_TYPE_NAME }?.let { walletType ->
            mutableTypes.remove(walletType)
            mutableTypes.add(walletType)
        }

        newTypeComboBox.items.addAll(
            mutableTypes.filter { it.name != Constants.GOAL_DEFAULT_WALLET_TYPE_NAME },
        )
    }

    private fun configureComboBoxes() {
        UIUtils.configureComboBox(walletComboBox) { it.name }
        UIUtils.configureComboBox(newTypeComboBox) { UIUtils.translateWalletType(it) }
    }

    private fun updateCurrentTypeLabel(wt: Wallet?) {
        if (wt == null) {
            currentTypeLabel.text = "-"
            return
        }
        currentTypeLabel.text = UIUtils.translateWalletType(wt.type)
    }
}
